### DISCUSSION TITLE
Some questions about the use of pump components and gas properties

### DISCUSSION THREAD

**Discussion by LGQ199810** [idaholab/moose#29053] [Q&A Modules: Thermal Hydraulics]
Hello, I am currently conducting calculations for a heat exchange circuit based on thermal hydraulic modules (the node diagram of the circuit is shown in the attached figure). During the calculation process, I found that the temperature difference between the pump components before and after exceeded a reasonable range, and the temperature difference between the pump components can reach about 50 ℃. Is there any adjustment method to reduce or even eliminate this difference.
In addition, in my previous calculations, when I used a custom gas property code, the cooling capacity of the cooling system col_sys was not as high as expected. After inspection, it was found that the reason was that the thermal conductivity calculated through my own code (about 0.033W/m · K) was lower than the constant value given by the ideal gas property package (about 0.041W/m · K). However, the thermal conductivity calculated by my own developed property code matched well with the properties in the REFPROP property package. Is there any way to solve this problem.
![1731327508775](https://github.com/user-attachments/assets/4393ce20-a4d8-435c-ba96-f94f145691f4)
the pump component used in my model
` [pump]

    type = Pump1Phase
    position = '-${col_connect} 0 -${fparse 2 * col_length2}'
    connections = 'col_sys/gcol_connect:out up_pipe:in'
    volume = 1e-4
    A_ref = '${fparse pi * col_gas_dia * col_gas_dia / 4.}'
    head = 5.
  []
`

**Comment by GiudGiud** [comment]
Which fluid properties are you currently using?

@joshuahansel

**Reply by LGQ199810** [reply]
The working fluid gas I am currently using is argon，and the input card I'm using as followed
[pump_test-upload.txt](https://github.com/user-attachments/files/17718447/pump_test-upload.txt)

**Comment by joshuahansel** [comment]
You mention "pumps". I see only one pump in your diagram. Do you just mean the difference across the one pump? Have you checked that the pressure change across your pump matches your supplied head value of 5 m? Remember that `dp = density*g*head`. Other pump formulation options for you are:

- https://mooseframework.inl.gov/source/components/ShaftConnectedPump1Phase.html
- https://mooseframework.inl.gov/source/components/ShaftConnectedCompressor1Phase.html

but these will be a bit more involved because you must specify performance curves and include a `Shaft` component (and other component(s)) on the shaft to power the pump. The one you are using is the simplest one, if you just want to match a certain pressure rise.

Regarding your thermal conductivity, you specified the value of 0.04071 W/(m-K) in your input file - you can change `k` to your desired value of 0.033 W/(m-K). That equation of state corresponds to an ideal gas formulation, which says nothing about transport properties like conductivity and viscosity, so the user must supply those, and they cannot have temperature dependence (though developers could change that part at some point). As far as I know, we do not have any specialized properties for argon, so you would need to implement your own FluidProperties class if you want something more accurate than the ideal gas equation of state.

**Reply by LGQ199810** [reply]
Thank you very much for providing the verification idea. Currently, I only use a single pump component to drive the gas in the circuit, so that the gas can flow and exchange heat in the circuit at a specific flow rate. I have not considered the shaft and other components. This is similar to the setting of pump components in the single_phase_flow chapter of the thermal hydraulic module tutorial. However, in my current circuit, the temperature difference between the pump components can reach about 50 ℃ (the temperature of col_stys/gcol_comnect: out is about 300K, and the temperature of uo_pipe: in is about 346K), which is contrary to common sense.
In addition, I have calculated the pressure difference before and after the pump, which is about 9.35e4 Pa. However, at this time, I have set the pump head to 0.0 m. There are also some issues with this calculation result that need to be corrected, so I would like to ask how to correct these two problems. Thank you very much.
![1731470311750](https://github.com/user-attachments/assets/e7b73813-0b3e-46ec-bd05-a19e42182577)

**Reply by joshuahansel** [reply]
You say

> However, at this time, I have set the pump head to 0.0 m

I don't know if you've changed your input file since, but in the one you attached, you're using a PID control to control the head parameter of the pump. And it appears that the PID controller is functioning correctly, and you've achieved the target mass flow rate. I ran your input, and it seems the head input signal to the pump is 850 m. For the pressure rise that you have across the pump, I am not convinced that a ~50 K difference is unreasonable. The pump pressure rise does not match this head (although it's quite a bit closer), but this is a compressible fluid, so I wouldn't necessarily expect it.

**Reply by LGQ199810** [reply]
I am very sorry for not stating in the previous question that I changed the head parameter filled in the pump component in the attachment input card from 5.0 to 0.0, but this seems to have no impact on the calculation results.
Meanwhile, in the previous calculations, I set the parameters of the pump to be completely consistent with the PID control of the pump in the single_phase_flow chapter of the thermal hydraulic module tutorial, and found that the pump components needed to be calculated for over 10000 seconds to achieve the desired loop flow rate. Is it necessary to adjust the PID parameters accordingly because the pressure and working fluid of the circuit I calculated are different from those in the tutorial.
In addition, in the circuit I am currently calculating, how to reduce the pressure head value of the pump to reduce the pressure changes at the inlet and outlet of the pump? I just want to achieve forced circulation of gas in the circuit, and do not need the boost provided by the pump; Also, if possible, could you teach me how to display the real-time head parameter of the pump in the output? Thank you very much.

**Reply by joshuahansel** [reply]
> Is it necessary to adjust the PID parameters

I recommend that you select your own. Unfortunately, it's a bit of trial and error. The coefficients in that tutorial may not work well for your situation. By the way, I usually set the derivative coefficient to 0.

> how to reduce the pressure head value of the pump to reduce the pressure changes at the inlet and outlet of the pump?

For a given mass flow rate, you would need to examine the pressure drops in the rest of the system. The pump head will need to counteract those.

> I just want to achieve forced circulation of gas in the circuit, and do not need the boost provided by the pump

I don't understand. To me, "forced circulation" means that you require the boost provided by the pump. Otherwise you rely on natural convection effects.

> how to display the real-time head parameter of the pump in the output?

```
  [pump_head]
    type = RealControlDataValuePostprocessor
    control_data_name = pid:output
  []
```

**Reply by LGQ199810** [reply]
Thank you very much for your previous answer. I have indeed fallen into a common sense misconception, which is to consider driving gas flow and boosting gas separately. I have checked the main pressure drop in my circuit and would like to ask you the following questions:

1. In my circuit, the flow area of the pipes is not consistent, and there may be a situation as shown in the figure below (this is a part of my calculation of the circuit, which is similar to the previous node diagram and uses a pump for driving), where the middle pipe is thin (with a diameter of 16mm) and the two sides of the pipe are thick (with a diameter of 30mm). However, the pressure drop in these areas is very obvious. Is there any method or special component (such as a tapered pipe or a tapered pipe) that can minimize the pressure drop of nearly 20.0 kPa caused by sudden changes in area as much as possible;
![1731588133000](https://github.com/user-attachments/assets/ccd74075-e82d-43e7-9163-c0478ef6b19a)

2. In addition, in the above model, when I expand the area of inlet_pipe and outlet_pipe to the flow area of down_pipe, the calculation will not converge due to being less than the minimum time step, and the calculation failure time is mostly around 0.0006s. What strategies do you have to adjust the input card to make the calculation converge for this type of calculation failure; Meanwhile, when I only change the area of one of inlet_pipe and outlet_pipe, the calculation can still converge, but the location where a significant pressure drop occurs will shift from jct3 to jct1;

3. Besides, when reading the temperature and pressure of each pipeline in the circuit, I found that for down_pipe, gas flowing through this area will cause a slight temperature rise (from 622K to 624K without heating). I speculate that this is due to friction and other reasons. However, there is no such phenomenon in outlet_pipe with the same input parameters. May I ask about the possible reasons for this phenomenon and if there are any solutions; Currently, I have not specified the wall roughness for any pipeline (the input card for down_pipe is as follows)

```  
[down_pipe_3]
    type = FlowChannel1Phase
    position = '${fparse connect_length * 2 + exp_length} 0 0'
    orientation = '0 0 -1'
    length = ${down_pipe_length}
    n_elems = ${down_pipe_n_elems}
    A = '${fparse pi * down_pipe_dia * down_pipe_dia / 4.}'
    D_h = ${down_pipe_dia}
    initial_p = 1.0e6
    initial_T = 600.0
    initial_vel = 3.0
    fp = ar
  []
```

**Reply by joshuahansel** [reply]
1. `FlowChannel1Phase` can be tapered to your liking; the parameter `A` takes the name of a `Function`, so you can make one that linear interpolates the area from A1 to A2.
2. This is a general convergence criteria, so there's no specific recommendation. Smoothing out discontinuities (both in space and time) is helpful. Otherwise you may want to play with tolerances and scaling factors to make sure you're not trying to over-solve. I typically use `scaling_factor_1phase = '1 1 1e-5'` for `FlowChannel1Phase`, `scaling_factor_rhoEV = 1e-5` for the junction/junction-like (e.g., pump) components, `scaling_factor_temperature = 1e-5` for the heat structures.
3. Friction would cause it, as well as the area change at `jct3` since that'll compress it.

**Reply by LGQ199810** [reply]
Thank you for your answer. Based on your answer to question 1, when I attempting to interpolate the area of Flowchannel1Phase using a function, the program reported an error: 'The desired parameter' component/gas_inlet_pipe/A 'was not located for the' gas_inlet_pipe_area_ctl 'object, It either does not exist or has not been declared as controllable. 'If possible, could you provide a relatively correct usage method.
My initial goal for this function is to reduce the diameter of the gas_inlet_pipe component from down_pipe_dia (i.e. 30mm) to exp_gas_dia (i.e. 16mm) within the range of 0.0m to 0.3m on the component. I referred to the definition method of the secondary side flow rate in the single_phase_flow chapter of the thermal hydraulic module tutorial to write the control module and function. The main input cards are as follows:
```
exp_gas_dia = '${units 16.0 mm -> m}'
down_pipe_dia = '${units 30.0 mm -> m}'

[Functions]
  [area_ctl]
    type = PiecewiseLinear
    xy_data = '
      0.0   ${down_pipe_dia}
      0.3   ${exp_gas_dia}'
  []
[]

[Components]
  [gas_inlet_pipe]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '1 0 0'
    length = ${connect_length}
    n_elems = ${connect_n_elems}
    A = ${fparse pi * exp_gas_dia * exp_gas_dia / 4.}'
    D_h = ${exp_gast_dia}
    initial_p = 1.0e6
    initial_T = 1100.0
    initial_vel = 3.0
    fp = ar
  []
[]

[ControlLogic]
  [gas_inlet_pipe_area]
    type = GetFunctionValueControl
    function = area_ctl
  []

  [gas_inlet_pipe_area_ctl]
    type = SetComponentRealValueControl
    component = gas_inlet_pipe
    parameter = A
    value = gas_inlet_pipe_area:value
  []
[]
```

**Reply by joshuahansel** [reply]
Delete the `gas_inlet_pipe_area_ctl` block and replace `A = ${fparse pi * exp_gas_dia * exp_gas_dia / 4.}'` with `A = area_ctl`.

There are some parameters that accept `Function` names directly. Others just take a floating point value, and some of these can be controlled. Check the documentation page for the class if you're ever unsure - it has tables of the input parameters and their types and whether they're controllable.
